CAF-2613 Tenant index field type mappings

CJComac opened the pull request.

mulhollandc committed:
Cherry picked the CAF-2758-code-review-changes branch to get the ElasticAuditConstants changes

Commit:
Type mappings reside as a resource, the resource is read in as a stream, parsed as JSON, added to mapping builder and then added to the request to create the index

Commit:
Moved retry elastic search operation into a class. Moved delete index to the Generated Audit Log integration test

Commit:
Added verify type mappings to ElasticAuditIT testESIndexing() test. Performed and removed CAF-2613 TODO comments.

glucy commented:
@CJComac There are currently conflicts.  Please resolve.  I will review in due course.

glucy commented on caf-audit/src/main/java/com/hpe/caf/auditing/elastic/ElasticAuditIndexManager.java:
```diff
@@ -18,12 +18,18 @@
+
+import java.io.IOException;
+
+import static org.elasticsearch.common.xcontent.XContentFactory.jsonBuilder;
```
Organise imports.
Remove * import, use explicits

CJComac replied:
Cleaned up the imports in the latest commit

glucy commented on caf-audit/src/main/java/com/hpe/caf/auditing/elastic/ElasticAuditIndexManager.java:
```diff
@@ -80,6 +89,34 @@ private void createIndex(String indexName){
+        } catch (IOException e) {
+            String errorMessage = "Unable to read bytes from " + cafAuditEventTenantIndexMappingsFileName;
+            LOG.error(errorMessage);
+            throw new RuntimeException(e);
```
Add message and re-throw exception.

CJComac replied:
Adjusted in the latest commit

Commit:
Cleaned up imports and adding error message to the RuntimeExceptions that are thrown

Commit:
Merge branch 'develop' into CAF-2613_Tenant_index_field_type_mappings

CJComac commented:
@glucy I made your feedback changes and merged develop into the branch

glucy commented on caf-audit/src/main/java/com/hpe/caf/auditing/elastic/ElasticAuditIndexManager.java:
```diff
@@ -33,6 +41,8 @@
     private final LoadingCache<String, String> indexCache;
     private final ElasticAuditConfiguration config;
 
+    private static final String indexTypeName = "cafAuditEvent";
```
Add this to the constants file class

CJComac replied:
Using the value in the constants file that holds the index type name in the latest commit.

glucy commented on caf-audit/src/main/java/com/hpe/caf/auditing/elastic/ElasticAuditIndexManager.java:
```diff
@@ -80,6 +91,34 @@ private void createIndex(String indexName){
                 .build();
         CreateIndexRequest indexRequest = new CreateIndexRequest(indexName, indexSettings);
 
+        String cafAuditEventTenantIndexMappingsFileName = "CafAuditEventTenantIndexMappings.json";
```
Place "CafAuditEventTenantIndexMappings.json" in the constants file also.

CJComac replied:
Moved into constants file in the latest commit.

glucy commented on caf-audit/src/test/java/com/hpe/caf/auditing/elastic/ElasticAuditIT.java:
```diff
@@ -19,10 +19,12 @@
+import org.elasticsearch.common.compress.CompressedXContent;
 import org.elasticsearch.index.query.QueryBuilders;
 import org.elasticsearch.rest.RestStatus;
 import org.elasticsearch.search.SearchHit;
```
Replace import java.util.*; with individual import.
You should do this directly in develop and pull it into your branch.

We should always look to improve :-)

CJComac replied:
Committed to develop and merged develop into this branch

glucy commented on caf-audit/src/test/java/com/hpe/caf/auditing/elastic/GeneratedAuditLogIT.java:
```diff
@@ -38,9 +39,12 @@
 
+    private static final String testTenant = "tenant1";
+
     private static final Logger LOG = LoggerFactory.getLogger(GeneratedAuditLogIT.class);
```
LOG is never used.  I suggest we use it.
If not used it should be removed, but I'd rather see it used.  @CJComac use it where appropriate in your tests, you don't need to retro-fit it.

CJComac replied:
Removed in the latest commit. Logger is not used in ElasticAuditIT.

mulhollandc commented on caf-audit/src/test/java/com/hpe/caf/auditing/elastic/RetryElasticsearchOperation.java:
```diff
@@ -0,0 +1,67 @@
+ */
+package com.hpe.caf.auditing.elastic;
+
+public class RetryElasticsearchOperation
```
Should this be renamed to ElasticAuditRetryOperation to be consistent with other classes under elastic?

CJComac replied:
Renamed in the latest commit.

Commit:
Merge branch 'develop' into CAF-2613_Tenant_index_field_type_mappings

Commit:
Merge branch 'develop' into CAF-2613_Tenant_index_field_type_mappings

Commit:
- Using the value in the constants file that holds the index type name
- Moved name of resource file, holding the type definition, into constants
- Removed logger from ElasticAuditIT
- Renamed RetryElasticsearchOperation to ElasticAuditRetryOperation

Commit:
Moved loading and parsing of type mappings JSON into class constructor

Commit:
Move construction of the index type mappings builder into a return method used by the constructor

Commit:
Get of the tenant index type mappings builder happens after the check to see if the tenant index exists and the mapping builder has not been set before.
